✨ Add PlacementGroupName and PlacementGroupPartition to AWSLaunchTemplate

justinmir opened the pull request:
/kind feature

**What this PR does / why we need it**:

This change adds support for specifying placement group configuration in AWSMachinePool launch templates.

New fields added to AWSLaunchTemplate:
- PlacementGroupName: specifies the name of the placement group in which to launch instances
- PlacementGroupPartition: specifies the partition number within a partition placement group (valid values: 1-7)

This enables users to target specific partitions when using partition placement groups with MachinePool resources, providing higher levels of availability by ensuring instances in different partitions do not share underlying hardware.

**Which issue(s) this PR fixes**
Relates-to: #4870

**Special notes for your reviewer**:
This change mirrors https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4870 but for launch templates. This allows us to create ASGs that target a specific placement group partition.

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/placement-groups.html
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Placement.html

**Checklist**:
- [x] squashed commits
- [x] includes documentation
- [x] includes emoji in title 
- [x] adds unit tests
- [x] adds or updates e2e tests

**Release note**:
```release-note
Adds support for specifying placement group configuration for AWSMachinePool launch templates.
```

justinmir committed:
Add PlacementGroupName and PlacementGroupPartition to AWSLaunchTemplate

This change adds support for specifying placement group configuration
in AWSMachinePool launch templates.

New fields added to AWSLaunchTemplate:
- PlacementGroupName: specifies the name of the placement group in
  which to launch instances
- PlacementGroupPartition: specifies the partition number within a
  partition placement group (valid values: 1-7)

This enables users to target specific partitions when using partition
placement groups with MachinePool resources, providing higher levels
of availability by ensuring instances in different partitions do not
share underlying hardware.

Relates-to: #4870

AndiDog reviewed (commented):
/ok-to-test

I didn't have a detailed look, but I'm pretty sure this needs more coverage, for example in `LaunchTemplateNeedsUpdate` (detect changes) and webhooks (to catch invalid combinations).

justinmir committed:
Add placement group coverage in LaunchTemplateNeedsUpdate and webhook

- LaunchTemplateNeedsUpdate now detects changes to PlacementGroupName
  and PlacementGroupPartition, triggering a new launch template version
  when either field changes
- AWSMachinePool webhook validates that PlacementGroupPartition is not
  set without PlacementGroupName, catching the invalid combination at
  admission time rather than during reconciliation

justinmir committed:
fix tests
